File: lib/results_screen.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:ningapi/app_colors.dart';
import 'package:ningapi/models/recognition_result.dart';
import 'package:ningapi/services/tts_service.dart';
import 'package:image_picker/image_picker.dart';
import 'package:ningapi/scanning_screen.dart';

class ResultsScreen extends StatefulWidget {
  final File imageFile;
  final CurrencyResult currencyResult;

  const ResultsScreen({
    super.key,
    required this.imageFile,
    required this.currencyResult,
  });

  @override
  State<ResultsScreen> createState() => _ResultsScreenState();
}

class _ResultsScreenState extends State<ResultsScreen> {
  final TtsService _tts = TtsService();
  final ImagePicker _picker = ImagePicker();
  bool _isProcessing = false;
  bool _isSpeaking = false;
  bool _isDisposed = false;

  @override
  void initState() {
    super.initState();
    // Delai pour laisser la navigation se terminer (augmente a 600ms pour eviter conflit avec dispose)
    Future.delayed(const Duration(milliseconds: 600), () {
      if (!_isDisposed && mounted) {
        _announceResult();
      }
    });
  }

  @override
  void dispose() {
    _isDisposed = true;
    _tts.stop();
    super.dispose();
  }

  Future<void> _announceResult() async {
    if (_isSpeaking || _isDisposed) return;
    _isSpeaking = true;

    try {
      // Annoncer le resultat et attendre la fin
      await _tts.speak(widget.currencyResult.speakableResult);

      if (_isDisposed || !mounted) return;

      // Attendre un peu puis annoncer l'aide (800ms est suffisant car speak attend deja la fin)
      await Future.delayed(const Duration(milliseconds: 800));

      if (_isDisposed || !mounted) return;

      await _tts.speak("Appuyez pour répéter. Balayez pour scanner un autre billet.");
    } finally {
      if (!_isDisposed) {
        _isSpeaking = false;
      }
    }
  }

  Future<void> _repeatResult() async {
    if (_isProcessing || _isSpeaking || _isDisposed) return;

    HapticFeedback.lightImpact();
    _isSpeaking = true;

    try {
      await _tts.speak(widget.currencyResult.speakableResult);
    } finally {
      if (!_isDisposed) {
        _isSpeaking = false;
      }
    }
  }

  Future<void> _scanAgain() async {
    if (_isProcessing || _isDisposed) {
      debugPrint("Scan deja en cours, swipe ignore");
      return;
    }

    _isProcessing = true;
    if (mounted) setState(() {});

    HapticFeedback.mediumImpact();

    try {
      // Arreter le TTS actuel
      await _tts.stop();

      // Utiliser speakAsync car on ne veut pas attendre
      await _tts.speakAsync("Ouverture de la camera");

      // Petit delai pour laisser le TTS commencer
      await Future.delayed(const Duration(milliseconds: 300));

      if (!mounted || _isDisposed) {
        _isProcessing = false;
        return;
      }

      // Ouvrir directement la camera
      final XFile? pickedFile = await _picker.pickImage(
        source: ImageSource.camera,
        preferredCameraDevice: CameraDevice.rear,
      );

      if (!mounted || _isDisposed) {
        _isProcessing = false;
        return;
      }

      if (pickedFile != null) {
        // Remplacer l'ecran actuel par ScanningScreen
        Navigator.of(context).pushReplacement(
          MaterialPageRoute(
            builder: (_) => ScanningScreen(imageFile: File(pickedFile.path)),
          ),
        );
      } else {
        // Utilisateur a annule, rester sur l'ecran des resultats
        _isProcessing = false;
        if (mounted && !_isDisposed) {
          setState(() {});
          await _tts.speak("Scan annule. Balayez pour reessayer.");
        }
      }
    } catch (e) {
      debugPrint("Erreur lors du scan: $e");
      _isProcessing = false;
      if (mounted && !_isDisposed) {
        setState(() {});
        await _tts.speak("Erreur camera. Veuillez reessayer.");
      }
    }
  }

  void _goBack() {
    // Retour a l'ecran d'accueil
    Navigator.of(context).popUntil((route) => route.isFirst);
  }

  @override
  Widget build(BuildContext context) {
    final resultText = widget.currencyResult.isUnknown 
        ? "Inconnu" 
        : widget.currencyResult.denomination;

    final resultSubtext = widget.currencyResult.isUnknown
        ? "Non reconnu"
        : (widget.currencyResult.currency == 'USD' ? "Dollars" : "Francs Congolais");
    
    final color = widget.currencyResult.isUnknown ? Colors.orange : AppColors.accent;

    return PopScope(
      canPop: false,
      onPopInvokedWithResult: (bool didPop, dynamic result) {
        if (didPop) return;
        _goBack();
      },
      child: Scaffold(
        backgroundColor: AppColors.backgroundLight,
        body: AbsorbPointer(
          absorbing: _isProcessing,
          child: GestureDetector(
            onTap: _repeatResult,
            onVerticalDragEnd: (details) {
              if (!_isProcessing) _scanAgain();
            },
            onHorizontalDragEnd: (details) {
              if (!_isProcessing) _scanAgain();
            },
            child: Semantics(
              label: "$resultText. $resultSubtext",
              hint: "Appuyez pour répéter. Balayez pour scanner un autre billet.",
              button: true,
              child: Container(
                color: Colors.transparent,
                width: double.infinity,
                height: double.infinity,
                child: SafeArea(
                  child: Stack(
                    children: [
                      Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          const Spacer(),

                          // Huge Status Icon
                          Icon(
                            widget.currencyResult.isUnknown
                              ? Icons.help_outline
                              : Icons.check_circle_outline,
                            size: 120,
                            color: color,
                          ),

                          const SizedBox(height: 40),

                          // Huge Result Text
                          Padding(
                            padding: const EdgeInsets.symmetric(horizontal: 20),
                            child: FittedBox(
                              fit: BoxFit.scaleDown,
                              child: Text(
                                resultText,
                                style: TextStyle(
                                  fontSize: 80,
                                  fontWeight: FontWeight.bold,
                                  color: AppColors.textDark,
                                ),
                              ),
                            ),
                          ),

                          const SizedBox(height: 10),

                          Text(
                            resultSubtext,
                            style: TextStyle(
                              fontSize: 30,
                              color: AppColors.secondary,
                              fontWeight: FontWeight.w600,
                            ),
                          ),

                          const Spacer(),

                          // Bottom Hint Area
                          Container(
                            width: double.infinity,
                            padding: const EdgeInsets.all(30),
                            color: AppColors.backgroundCard,
                            child: Column(
                              children: [
                                Icon(Icons.touch_app, size: 40, color: AppColors.textGray.withOpacity(0.5)),
                                const SizedBox(height: 10),
                                const Text(
                                  "Appuyer pour Répéter",
                                  style: TextStyle(fontSize: 20, color: AppColors.textGray),
                                ),
                                const SizedBox(height: 20),
                                Icon(Icons.swipe, size: 40, color: AppColors.textGray.withOpacity(0.5)),
                                const SizedBox(height: 10),
                                const Text(
                                  "Balayer pour Scanner",
                                  style: TextStyle(fontSize: 20, color: AppColors.textGray),
                                ),
                              ],
                            ),
                          ),
                        ],
                      ),

                      // Loading overlay when processing
                      if (_isProcessing)
                        Container(
                          color: Colors.black.withOpacity(0.3),
                          child: const Center(
                            child: CircularProgressIndicator(
                              color: AppColors.accent,
                            ),
                          ),
                        ),
                    ],
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

File: .OTLIN/ui/screens/ResultsScreen.kt
package com.ningapi.ui.screens

import android.widget.Toast
import androidx.activity.compose.BackHandler
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectDragGestures
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircleOutline
import androidx.compose.material.icons.filled.HelpOutline
import androidx.compose.material.icons.filled.Swipe
import androidx.compose.material.icons.filled.TouchApp
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.FileProvider
import androidx.navigation.NavController
import com.ningapi.models.CurrencyResult
import com.ningapi.services.TtsService
import com.ningapi.ui.theme.AppColors
import kotlinx.coroutines.delay
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

@Composable
fun ResultsScreen(
    navController: NavController,
    imageFile: File,
    currencyResult: CurrencyResult
) {
    val context = LocalContext.current
    val ttsService = remember { TtsService.getInstance(context) }
    
    var isProcessing by remember { mutableStateOf(false) }

    // TTS Announcement
    LaunchedEffect(Unit) {
        delay(300)
        ttsService.speak(currencyResult.speakableResult)
        delay(2000)
        ttsService.speak("Appuyez pour repeter. Balayez pour scanner un autre billet.")
    }

    // Camera Logic for "Scan Again"
    var currentPhotoPath by remember { mutableStateOf<String?>(null) }
    
    fun createImageFile(): File? {
        val timeStamp: String = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
        val storageDir: File? = context.externalCacheDir
        return try {
            File.createTempFile("JPEG_${timeStamp}_", ".jpg", storageDir).apply {
                currentPhotoPath = absolutePath
            }
        } catch (ex: IOException) {
            null
        }
    }

    val cameraLauncher = rememberLauncherForActivityResult(
        contract = ActivityResultContracts.TakePicture()
    ) { success ->
        isProcessing = false
        if (success && currentPhotoPath != null) {
            navController.navigate("scanning/${currentPhotoPath}") {
                popUpTo("results") { inclusive = true }
            }
        } else {
             ttsService.speakAsync("Scan annule. Balayez pour reessayer.")
        }
    }

    fun scanAgain() {
        if (isProcessing) return
        isProcessing = true
        ttsService.speakAsync("Ouverture de la camera")
        
        val photoFile = createImageFile()
        if (photoFile != null) {
             val uri = FileProvider.getUriForFile(
                context,
                "${context.packageName}.fileprovider",
                photoFile
            )
            cameraLauncher.launch(uri)
        } else {
            isProcessing = false
            Toast.makeText(context, "Erreur création fichier", Toast.LENGTH_SHORT).show()
        }
    }

    BackHandler {
        navController.popBackStack("home", inclusive = false)
    }

    val resultText = if (currencyResult.isUnknown) "Inconnu" else currencyResult.denomination
    val resultSubtext = if (currencyResult.isUnknown) "Non reconnu" else if (currencyResult.currency == "USD") "Dollars" else "Francs Congolais"
    val color = if (currencyResult.isUnknown) Color(0xFFFF9800) else AppColors.Accent // Orange or Accent

    Scaffold(
        containerColor = AppColors.BackgroundLight
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .pointerInput(Unit) {
                    detectDragGestures(
                        onDragEnd = { scanAgain() },
                        onDrag = { _, _ -> } 
                    )
                }
                .pointerInput(Unit) {
                     detectTapGestures(
                         onTap = { ttsService.speakAsync(currencyResult.speakableResult) }
                     )
                }
        ) {
            Column(
                modifier = Modifier.fillMaxSize(),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                 Spacer(modifier = Modifier.weight(1f))

                // Icon
                Icon(
                    imageVector = if (currencyResult.isUnknown) Icons.Default.HelpOutline else Icons.Default.CheckCircleOutline,
                    contentDescription = null,
                    modifier = Modifier.size(120.dp),
                    tint = color
                )

                Spacer(modifier = Modifier.height(40.dp))

                // Result Text
                Text(
                    text = resultText,
                    style = MaterialTheme.typography.displayLarge.copy(
                        fontWeight = FontWeight.Bold,
                        color = AppColors.TextDark,
                        fontSize = 80.sp
                    ),
                    textAlign = TextAlign.Center,
                    maxLines = 1
                )

                Spacer(modifier = Modifier.height(10.dp))

                Text(
                    text = resultSubtext,
                    style = MaterialTheme.typography.headlineMedium.copy(
                         color = AppColors.Secondary,
                         fontWeight = FontWeight.SemiBold,
                         fontSize = 30.sp
                    ),
                    textAlign = TextAlign.Center
                )

                Spacer(modifier = Modifier.weight(1f))

                // Bottom Hint
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .background(AppColors.BackgroundCard)
                        .padding(30.dp)
                ) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally, modifier = Modifier.fillMaxWidth()) {
                        Icon(Icons.Default.TouchApp, contentDescription = null, modifier = Modifier.size(40.dp), tint = AppColors.TextGray.copy(alpha = 0.5f))
                        Spacer(modifier = Modifier.height(10.dp))
                        Text("Appuyer pour Répéter", style = TextStyle(fontSize = 20.sp, color = AppColors.TextGray))
                        
                        Spacer(modifier = Modifier.height(20.dp))
                        
                        Icon(Icons.Default.Swipe, contentDescription = null, modifier = Modifier.size(40.dp), tint = AppColors.TextGray.copy(alpha = 0.5f))
                        Spacer(modifier = Modifier.height(10.dp))
                        Text("Balayer pour Scanner", style = TextStyle(fontSize = 20.sp, color = AppColors.TextGray))
                    }
                }
            }

            if (isProcessing) {
                Box(
                    modifier = Modifier
                        .fillMaxSize()
                        .background(Color.Black.copy(alpha = 0.3f)),
                    contentAlignment = Alignment.Center
                ) {
                    CircularProgressIndicator(color = AppColors.Accent)
                }
            }
        }
    }
}
